File: server/src/main/java/com/example/server/note/controller/NoteController.java
package com.example.server.note.controller;

import com.example.server.note.domain.Markdown;
import com.example.server.note.service.MarkdownServiceImpl;
import com.example.server.note.service.NoteServiceImpl;
import com.example.server.note.domain.Note;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.http.ResponseEntity;

import java.util.List;
import java.util.Optional;

@RestController
@RequiredArgsConstructor
public class NoteController {
    private final NoteServiceImpl noteService;
    private final MarkdownServiceImpl markdownService;

    @GetMapping("/notes")
    public List<Note> getNotes() {
        return noteService.getNotes();
    }

    @GetMapping("/notes/{slug}")
    public ResponseEntity<Note> getNote(@PathVariable String slug) {
        return noteService.getNote(slug)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PostMapping("/notes")
    public Note postNote(@RequestBody Note note) {
        Markdown md = new Markdown();
        md.setSlug(note.getSlug());
        md.setContent("");
        markdownService.saveMarkdown(md);
        return noteService.saveNote(note);
    }
}

File: server/src/main/kotlin/work/notespace/server/note/controller/NoteController.kt
package work.notespace.server.note.controller

import jakarta.validation.Valid
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import work.notespace.server.note.dto.NoteDto
import work.notespace.server.note.service.NoteService

@RestController
@Validated
class NoteController(
    private val noteService: NoteService,
) {
    @GetMapping("/notes")
    fun getNotes(): List<NoteDto> {
        return noteService.getNotes()
    }

    @GetMapping("/notes/{slug}")
    fun getNote(
        @PathVariable slug: String
    ): NoteDto {
        return noteService.getNote(slug)
    }

    @PostMapping("/notes")
    fun createNote(
        @Valid
        @RequestBody note: NoteDto
    ): NoteDto {
        return noteService.createNote(note)
    }

    @GetMapping("/notes/recent-created")
    fun getRecentCreatedNotes(
        @RequestParam(defaultValue = "10") limit: Int
    ): List<NoteDto> {
        return noteService.getRecentCreatedNotes(limit)
    }

    @GetMapping("/notes/recent-updated")
    fun getRecentUpdatedNotes(
        @RequestParam(defaultValue = "10") limit: Int
    ): List<NoteDto> {
        return noteService.getRecentUpdatedNotes(limit)
    }

    @GetMapping("/notes/random")
    fun getRandomNote(): NoteDto {
        return noteService.getRandomNote()
    }
}
